akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessId.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.typed.internal
+
+import akka.annotation.InternalApi
+import akka.cluster.sharding.typed.ShardingEnvelope
+import akka.cluster.sharding.typed.ShardingMessageExtractor
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ShardedDaemonProcessId {
+
+  /*
+   * The entity id used for sharded daemon process has the format: [revision]|[total-count]|[this-process-id]
+   *
+   * Pingers repeatedly ping each worker id for a given number of processes to keep them alive. When rescaling
+   * to a new number of workers the revision is increased so that we can detect pings for old processes and avoid
+   * starting old workers anew.
+   */
+  val Separator = '|'
+
+  def decodeEntityId(id: String, supportsRescale: Boolean) =
+    if (supportsRescale) {
+      id.split(Separator) match {
+        case Array(rev, count, n) => DecodedId(rev.toLong, count.toInt, n.toInt)
+        case _                    => throw new IllegalArgumentException(s"Unexpected id for sharded daemon process: '$id'")
+      }
+    } else {
+      DecodedId(0L, 0, id.toInt) // for non-dynamic-scaling workers, total count not used
+    }
+
+  final case class DecodedId(revision: Long, totalCount: Int, processNumber: Int) {
+    def encodeEntityId(supportsRescale: Boolean): String =
+      if (supportsRescale) s"$revision$Separator$totalCount$Separator$processNumber"
+      else processNumber.toString
+  }
+
+  final class MessageExtractor[T](supportsRescale: Boolean) extends ShardingMessageExtractor[ShardingEnvelope[T], T] {
+    def entityId(message: ShardingEnvelope[T]): String = message match {
+      case ShardingEnvelope(id, _) => id
+    }
+
+    // use process n for shard id
+    def shardId(entityId: String): String = {
+      if (supportsRescale) entityId.split(Separator)(2)
+      else entityId
```

Comment:
Hmmm, the pingers live on the old nodes, so may send the single id ping until roll complete, but extractor on new nodes now that it supports rescale, but they can't get revision and total count from anywhere. 

Not sure what sharding does for a failing extractor, but we'd sort of want to drop those Start messages somehow when `supportsRescale` but getting a non-rescale message. Once the roll is completed the pingers will live on new nodes and send the right Start messages. 

Maybe we can parse such messages into revision -1 and the startup check will cancel starting the worker and we'll be fine.

Reply:
Something more around this: cluster singleton will not start pinging nodes until all nodes are rolled since it will run on oldest - unless you limit it with role, in that scenario there could still be old nodes left when all of a certain role has rolled up, so there is probably a caveat to document for rolling upgrades there.

Reply:
Went with just id in string for revision 0 in [b1998c8](https://github.com/akka/akka/pull/31844/commits/b1998c8dc6f33d8fcee737ef08b2adcc754a4c6a), should work, but didn't figure out how to add a multimode test covering it to be sure.